Title:	Unable to delete reading due to xref
Stat:	open
Disp:	
Prio:	high
Cats:	bug
Reqtrs:	
AssdTo:	stuart

2011-02-26 11:29:00 stuart
 Rene Malenfant and Jim Breen report on the edict-editors list on
 2011-02-25, "Entry for removal", that they are unable to remove
 the いっそ reading from entry q=1164340.

 This seems to be remenicient of IS-198, "edconf.py crash when
 deleting entry with rev xrefs".

2011-02-26 11:32:00 stuart
 When I try to duplicate action, submission results in edsubmit.py
 failing:

  Traceback (most recent call last):
  File "/home/stuart/jdb/jb/web/cgi/edsubmit.py", line 644, in <module>
    main (args, opts)
  File "/home/stuart/jdb/jb/web/cgi/edsubmit.py", line 183, in main
    sess.userid if sess else None)
  File "/home/stuart/jdb/jb/web/cgi/edsubmit.py", line 369, in submission
    added = approve (dbh, entr, edtree, errs)
  File "/home/stuart/jdb/jb/web/cgi/edsubmit.py", line 425, in approve
    res = addentr (dbh, entr)
  File "/home/stuart/jdb/jb/web/cgi/edsubmit.py", line 469, in addentr
    res = jdb.addentr (dbh, entr)
  File "/home/stuart/jdb/jb/python/lib/jdb.py", line 1404, in addentr
    for x in getattr (s, '_xrer',  []): dbinsert (cur, "xref",  ['entr','sens','xref','typ','xentr','xsens','rdng','kanj','notes'], x)
  File "/home/stuart/jdb/jb/python/lib/jdb.py", line 120, in dbinsert
    raise e
  IntegrityError: insert or update on table "xref" violates foreign key constraint "xref_rdng_fkey"
  DETAIL:  Key (xentr, rdng)=(1058895, 2) is not present in table "rdng".

 That is, when creating the new entry, it is trying to duplicate a
 existing reverse xref that references reading #2 (which we just 
 removed.)

 The entry being edited (id=1058864) shows an xref to q2613280 (which has 
 only one entry, id=1058886)
 Looking at the reverse xrefs for the entry being edited (i.e. those that
 have 'xentr' pointing to the edit entry) shows:

  jmdict=# select * from xref where xentr=1058864;
    entr   | sens | xref | typ |  xentr  | xsens | rdng | kanj | notes 
  ---------+------+------+-----+---------+-------+------+------+-------
   1058757 |    1 |    1 |   3 | 1058864 |     1 |    2 |      | 
   1058886 |    1 |    1 |   3 | 1058864 |     3 |    1 |    1 | 

 shows additional entry 1058757 which turns out to be a rejected entry
 that references reading #2 of the edit entry.

2011-02-26 21:17:00 stuart
 Although the problem manifests in edsumbit.py, its genisis is in 
 edconf.py where the reverese xref (_xrer) lists are copied from the
 parent entry to the new edited entry.
 Added a filter there to strip out any reverse xrefs that have reading
 or kanj numbers greater than the readngs or kanj available in the target
 entry.

 Note there is another problem unaddressed yet: xrefs reference target
 readings and kanji by ordinal position so that when readigns or kanji
 are reordered or deleted, the xref target reading or kanji will "shift"
 to a possibly unintended one.

